### PR TITLE
Add instance_template template

### PR DESCRIPTION
**Pull request by sourced-pavel**

### CODE REVIEW

File: community/cloud-foundation/templates/instance_template/instance_template.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                        'machineType': machine_type,
+                        'disks': [boot_disk],
+                        'networkInterfaces': [network]
+                    }
+            }
+        }
+
+    template_spec = instance_template['properties']['properties']
+
+    optional_props = ['metadata',
```
Comment (tiadobatima1) [inline community/cloud-foundation/templates/instance_template/instance_template.py]: minor: Did `yapf` format this list like this?

File: community/cloud-foundation/templates/instance_template/instance_template.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                      'tags',
+                      'canIpForward',
+                      'labels',
+                      'serviceAccounts']
+
+    for prop in optional_props:
+        set_optional_property(template_spec, properties, prop)
+
+    set_optional_property(template_spec,
```
Comment (tiadobatima1) [inline community/cloud-foundation/templates/instance_template/instance_template.py]: `yapf` weirdness again?

File: community/cloud-foundation/templates/instance_template/instance_template.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                          properties,
+                          'instanceDescription',
+                          'description')
+
+    set_optional_property(instance_template['properties'],
```
Comment (tiadobatima1) [inline community/cloud-foundation/templates/instance_template/instance_template.py]: `yapf`?

File: community/cloud-foundation/templates/instance_template/instance_template.py (hunk 1 above)
Comment (sourced-pavel) [reply community/cloud-foundation/templates/instance_template/instance_template.py]: nope, that was my initiative. Fixed

File: community/cloud-foundation/templates/instance_template/instance_template.py (hunk 2 above)
Comment (sourced-pavel) [reply community/cloud-foundation/templates/instance_template/instance_template.py]: Fixed

File: community/cloud-foundation/templates/instance_template/instance_template.py (hunk 3 above)
Comment (sourced-pavel) [reply community/cloud-foundation/templates/instance_template/instance_template.py]: Fixed

File: community/cloud-foundation/templates/instance_template/instance_template.py.schema [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        count: 3
+
+outputs:
+  properties:
+    - name:
+        type: string
+        description: The instance template name.
+    - selfLink:
+        type: string
+        description: The instance temlpate URL.
```
Comment (sourced-vince) [inline community/cloud-foundation/templates/instance_template/instance_template.py.schema]: Mind changing this to `The URI (SelfLink) of the instance template resource.` as i will be updating the other templates accordingly.

File: community/cloud-foundation/templates/instance_template/instance_template.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    return boot_disk
+
+
+def get_network(properties):
+    """ Gets configuration that connects an instance to an existing network
+        and assigns to it an ephemeral public IP.
+    """
+
+    network_name = properties.get('network')
+    is_self_link = '/' in network_name or '.' in network_name
```
[the remark this answers is not in the document]
Comment (sourced-pavel) [reply community/cloud-foundation/templates/instance_template/instance_template.py]: yes, if it's the reference

File: community/cloud-foundation/templates/instance_template/tests/integration/instance_template.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        items:
+          - ftp
+          - https
+      metadata:
+        items:
+          - key: createdBy
+            value: unitTest
+      labels:
+        name: wrench
+  - name: test-network-${RAND}
```
[the remark this answers is not in the document]
Comment (sourced-pavel) [reply community/cloud-foundation/templates/instance_template/tests/integration/instance_template.yaml]: Yes, it's created for test.
I actually already developed several templates with tests prerequisites built into the YAML. The reasoning is:
1) We don't use 2 different tools for creating related resources
2) Test cleanup is equal to deployment delete. There's no extra step which also has to succeed.
3) It's simpler to develop test this way. E.g. if I don't want to delete the deployment after test exists, I don't need to take care of test teardown as well.
4) Test YAML itself becomes one more (very useful) example, showing multiple dependent resources created from same configuration.

File: community/cloud-foundation/templates/instance_template/instance_template.py [hunk 7]
```diff
@@ -81,39 +83,46 @@ def generate_config(context):
                         'networkInterfaces': [network]
                     }
             }
-        }
+    }
 
     template_spec = instance_template['properties']['properties']
```
[the remark this answers is not in the document]
Comment (sourced-pavel) [reply community/cloud-foundation/templates/instance_template/instance_template.py]: That's legit property name, I swear:
```
properties:
  properties:
    ...
```

File: community/cloud-foundation/templates/instance_template/instance_template.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+            'type': 'ONE_TO_ONE_NAT'
+        }]
+    }
+
+
+def generate_config(context):
+    """ Entry point for the deployment resources. """
+
+    properties = context.properties
+    name = properties.get('name', context.env['name'] + '-it')
```
[the remark this answers is not in the document]
Comment (sourced-pavel) [reply community/cloud-foundation/templates/instance_template/instance_template.py]: None, actually. I'll remove it.

File: community/cloud-foundation/templates/instance_template/instance_template.py.schema (hunk 4 above)
Comment (sourced-pavel) [reply community/cloud-foundation/templates/instance_template/instance_template.py.schema]: Will do.

File: community/cloud-foundation/templates/instance_template/tests/integration/instance_template.yaml [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+# Variables:
+#   RAND: a random string used by the testing suite
+#   IMAGE: a URL to the base disk image provided by the testing suite
+
+imports:
+  - path: templates/instance_template/instance_template.py
+    name: instance_template.py
+
+resources:
+  - name: instane-template-${RAND}
```
[the remark this answers is not in the document]
Comment (sourced-pavel) [reply community/cloud-foundation/templates/instance_template/tests/integration/instance_template.yaml]: Fixed

File: community/cloud-foundation/templates/instance_template/instance_template.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (sourced-pavel) [reply community/cloud-foundation/templates/instance_template/instance_template.py]: Done

File: community/cloud-foundation/templates/instance_template/instance_template.py.schema (hunk 4 above)
Comment (sourced-pavel) [reply community/cloud-foundation/templates/instance_template/instance_template.py.schema]: Done

File: community/cloud-foundation/templates/instance_template/instance_template.py [hunk 10]
```diff
@@ -67,7 +67,7 @@ def generate_config(context):
     """ Entry point for the deployment resources. """
 
     properties = context.properties
-    name = properties.get('name', context.env['name'] + '-it')
+    name = properties.get('name') or context.env['name']
```
Comment (sourced-vince) [inline community/cloud-foundation/templates/instance_template/instance_template.py]: Ooops. Just one more thing. Could you do `properties.get('name', context.env['name'])

Comment (sourced-pavel) [reply community/cloud-foundation/templates/instance_template/instance_template.py]: Fixed

**Pull request closed without merging**